Convert tabs component to TypeScript

tarunvijwani opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

With this PR, we're converting the Tabs component to TypeScript

Fixes #7863

## Changes in the PR:

- Added types for the Tabs component. 
- Added Tabs component to Storybook. 


<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->


#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.




### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests


#### User Facing Testing
1. Checkout this branch
2. Run `npm run storybook`
3. Go Storybook-> Tabs ->Docs
4. Test the block. 

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

Commit:
Convert tabs component to TypeScript

Commit:
Add tabs component story

Commit:
Merge branch 'trunk' into update/7863-convert-tabs-to-TS

tarhi-saad commented on assets/js/base/components/tabs/index.tsx:
```diff
@@ -10,6 +10,22 @@ import { useTabState, Tab, TabList, TabPanel } from 'reakit/Tab';
+		ariaLabel?: string;
+	} >;
+	activeClass?: string;
+	initialTabName?: string | undefined;
```
@tarunvijwani, the `initialTabName` is marked as optional (might be undefined). So, we don't need to specify that it can be undefined.

```suggestion
	initialTabName?: string;
```

opr replied:
This is not true, because we have `exactOptionalPropertyTypes` enabled https://github.com/woocommerce/woocommerce-blocks/blob/d9f0aa8b8e26bbf50f109ac1f845b22ac8be208b/tsconfig.base.json#L10 we do need to define it.

tarhi-saad reviewed (changes requested):
Hi, @tarunvijwani! Thank you for working on this PR! I've left a comment below.

The Storybook doesn't seem being implemented well. I can see only two controls:

<img width="1025" alt="image" src="https://user-images.githubusercontent.com/14235870/209972395-7ea88057-db57-45fb-8bbb-530cd3f71f32.png">

As opposed to other examples, we can see more information about the component:

<img width="1023" alt="image" src="https://user-images.githubusercontent.com/14235870/209972614-4ab00569-2258-4e7e-9630-42d2749c1e61.png">

Commit:
Add desciptions in comments to the Tab props

Commit:
Rename and export tabs component to display descriptions in Storybook

Renamed Tabs component to __Tabs as a sign it should not be used directly. Since we were using withInstanceId as the default export and were not exporting the Tabs component, Storybook could not retrieve the description and default values. We have renamed and exported the component to make sure Storybook can pick up the descriptions and default values.

opr commented on assets/js/base/components/tabs/index.tsx:
```diff
@@ -10,7 +10,50 @@ import { useTabState, Tab, TabList, TabPanel } from 'reakit/Tab';
+/**
+ * Exporting the component for Storybook. Use the default export instead.
+ */
+export const __Tabs = ( {
```
I thought `__TabsWithoutInstanceId` would be a better name for this, it's a little more descriptive. What do you think?

tarunvijwani replied:
Yes, that sounds good! Update the code. Thank you! 👍

opr reviewed (commented):
Looking good! I just mentioned the name of the exported "raw" component to be more descriptive, do you think we should change it?

Commit:
Rename exported Tabs component

Changed the exported Tabs component from __Tabs to __TabsWithoutInstanceId to make it more descriptive.

opr reviewed (approved):
Thanks! 👏🏼  looks perfect now.
